src/mon/OSDMonitor.cc

```diff
@@ -7784,22 +7784,37 @@ int OSDMonitor::get_crush_rule(const string &rule_name,
   return 0;
 }
 
-int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, ostream *ss)
+int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, int crush_rule, ostream *ss)
 {
   auto max_pgs_per_osd = g_conf().get_val<uint64_t>("mon_max_pg_per_osd");
-  auto num_osds = std::max(osdmap.get_num_in_osds(), 3u);   // assume min cluster size 3
-  auto max_pgs = max_pgs_per_osd * num_osds;
   uint64_t projected = 0;
-  if (pool < 0) {
-    projected += pg_num * size;
-  }
-  for (const auto& i : osdmap.get_pools()) {
-    if (i.first == pool) {
-      projected += pg_num * size;
-    } else {
-      projected += i.second.get_pg_num_target() * i.second.get_size();
+  unsigned osd_num = 0;
+  CrushWrapper newcrush;
+  _get_pending_crush(newcrush);
+  set<int> roots;
+  newcrush.find_takes_by_rule(crush_rule, &roots);
+  if (!roots.empty()) {
+    mapping.update(osdmap);
```

Comment:
why update `mapping` here?

Reply:
The previous failure to pass the test was due to the fact that the mapping was not created when osdmap.get_max_osd() was called, making it impossible to get the correct osd information.

Reply:
we deliberately split the job of updating `mapping` into smaller job slices and let `OSDMonitor::mapper` and friends to do this in background, so that the monitor is not choked when handling a large amount of pg.

i don't think this problem is addressed here. as `OSDMapMapping::update()` process *all* pools in a single batch. this defeats the purpose of putting the mapping task to the background.

one way to address this is to check if the `mapping.get_epoch()` is new enough before calling `mapping.get_osd_acting_pgs(id)` and use a safe number as a fallback.

Reply:
#41009 is created to remove `OSDMapMapping::update()`.